pkg/mcp/sink/sink.go

```diff
@@ -0,0 +1,323 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sink
+
+import (
+	"io"
+	"sort"
+	"sync"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/types"
+	"github.com/gogo/status"
+	"google.golang.org/grpc/codes"
+
+	mcp "istio.io/api/mcp/v1alpha1"
+	"istio.io/istio/pkg/log"
+	"istio.io/istio/pkg/mcp/monitoring"
+)
+
+var scope = log.RegisterScope("mcp", "mcp debugging", 0)
+
+// Object contains a decoded versioned object with metadata received from the server.
+type Object struct {
+	TypeURL  string
+	Metadata *mcp.Metadata
+	Body     proto.Message
+}
+
+// changes is a collection of configuration objects of the same protobuf type.
+type Change struct {
+	Collection string
+
+	// List of resources to add/update. The interpretation of this field depends
+	// on the value of Incremental.
+	//
+	// When Incremental=True, the list only includes new/updated resources.
+	//
+	// When Incremental=False, the list includes the full list of resources.
+	// Any previously received resources not in this list should be deleted.
+	Objects []*Object
+
+	// List of deleted resources by name. The resource name corresponds to the
+	// resource's metadata name.
+	//
+	// Ignore when Incremental=false.
+	Removed []string
```

Comment:
Just curious why is this needed? if no update received for a resource within a time frame it is destined to be removed.